### ISSUE TITLE
Test Data Representation Mismatch & Incomplete Pipeline Serialization in `LLM-Model/`

### ISSUE THREAD

**Issue by codeCraft-Ritik** [anshul-dying/EthicalEye_EDAI_SEM_5-#2]
> **Severity:** High | **Component:** ML Benchmark & Serialization Pipeline | **Type:** Model Evaluation / Data Pipeline Bug

---

### 📋 Overview
In `LLM-Model/compare_models.py`, `LLM-Model/class.py`, and `LLM-Model/dp.py`, all ML classifiers are fitted on **TF-IDF float representations**, but are evaluated against **raw integer count vectors** during test prediction. Additionally, model serialization (`.joblib`) discards the fitted `TfidfTransformer`, breaking subsequent inference.

---

### 🔍 Root Cause Analysis

| File | Problematic Line | Description |
| :--- | :--- | :--- |
| `LLM-Model/compare_models.py` | Line 52 | Evaluates `clf.predict(count_vect.transform(X_test))` without `tfidf_transformer` |
| `LLM-Model/class.py` | Line 40, 44-45 | Ignores TF-IDF on test set & fails to serialize `tfidf_transformer` |
| `LLM-Model/dp.py` | Line 36, 48-49 | Omits TF-IDF transformation step on validation and serialization |

```python
# --- Current Code in compare_models.py ---
X_train_counts = count_vect.fit_transform(X_train)
tfidf_transformer = TfidfTransformer()
X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)

for clf in classifiers:
    clf.fit(X_train_tfidf, y_train)           # 1. Fitted on TF-IDF floats
    y_pred = clf.predict(count_vect.transform(X_test)) # 2. Evaluated on raw CountVectorizer integers!
```

---

### 💥 Impact Breakdown
* **Distorted Benchmark Metrics:** Accuracy scores and confusion matrices reported across all 7 algorithms (SVM, Random Forest, Naive Bayes, SGD, Logistic Regression) are scientifically invalid due to feature space mismatch.
* **Corrupted Inference Artifacts:** Downstream scripts that load `category_classifier.joblib` and `category_vectorizer.joblib` feed raw counts into a classifier that expects normalized TF-IDF features.

---

### 🛠️ Proposed Solution (Code Diff)

```diff
  X_train_counts = count_vect.fit_transform(X_train)
  tfidf_transformer = TfidfTransformer()
  X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)

+ # Transform test data into the same TF-IDF space
+ X_test_counts = count_vect.transform(X_test)
+ X_test_tfidf = tfidf_transformer.transform(X_test_counts)

  for clf in classifiers:
      clf.fit(X_train_tfidf, y_train)
-     y_pred = clf.predict(count_vect.transform(X_test))
+     y_pred = clf.predict(X_test_tfidf)
```

#### ✨ Modern Best Practice (Scikit-Learn Pipeline):
```python
from sklearn.pipeline import Pipeline
from joblib import dump

# Encapsulate preprocessing & model into a single self-contained pipeline
pipeline = Pipeline([
    ('vectorizer', CountVectorizer()),
    ('tfidf', TfidfTransformer()),
    ('classifier', MultinomialNB())
])

pipeline.fit(X_train, y_train)
y_pred = pipeline.predict(X_test)

# Saves vectorizer, TF-IDF weights, and classifier together
dump(pipeline, 'dark_pattern_classifier_pipeline.joblib')
```
